File: packages/evaluatorq/src/lib/evaluatorq.ts
import { Effect, pipe } from "effect";

import type { Orq } from "@orq-ai/node";

import { processDataPointEffect } from "./effects.js";
import {
  ProgressService,
  ProgressServiceLive,
  withProgress,
} from "./progress.js";
import { sendResultsToOrqEffect } from "./send-results.js";
import { displayResultsTableEffect } from "./table-display.js";
import {
  captureParentContext,
  flushTracing,
  generateRunId,
  initTracingIfNeeded,
  shutdownTracing,
  type TracingContext,
} from "./tracing/index.js";
import {
  setEvaluationAttributes,
  setJobNameAttribute,
  withEvaluationSpan,
  withJobSpan,
} from "./tracing/spans.js";
import type {
  DataPoint,
  EvaluatorParams,
  EvaluatorqResult,
  Job,
} from "./types.js";

async function setupOrqClient(apiKey: string) {
  try {
    const client = await import("@orq-ai/node");
    const serverURL = process.env.ORQ_BASE_URL || "https://my.orq.ai";

    return new client.Orq({ apiKey, serverURL });
  } catch (error: unknown) {
    const err = error as Error & { code?: string };
    if (
      err.code === "MODULE_NOT_FOUND" ||
      err.code === "ERR_MODULE_NOT_FOUND" ||
      err.message?.includes("Cannot find module")
    ) {
      throw new Error(
        "The @orq-ai/node package is not installed. To use dataset features, please install it:\n" +
          "  npm install @orq-ai/node\n" +
          "  # or\n" +
          "  yarn add @orq-ai/node\n" +
          "  # or\n" +
          "  bun add @orq-ai/node",
      );
    }
    throw new Error(`Failed to setup ORQ client: ${err.message || err}`);
  }
}

interface DataPointBatch {
  datapoints: DataPoint[];
  hasMore: boolean;
  batchNumber: number;
}

async function* fetchDatasetBatches(
  orqClient: Orq,
  datasetId: string,
  options?: { includeMessages?: boolean },
): AsyncGenerator<DataPointBatch> {
  let startingAfter: string | undefined;
  let batchNumber = 0;
  let hasYielded = false;

  try {
    while (true) {
      const response = await orqClient.datasets.listDatapoints({
        datasetId,
        limit: 50,
        startingAfter,
      });

      if (!response.data || response.data.length === 0) {
        if (!hasYielded) {
          throw new Error(`Dataset ${datasetId} not found or has no data`);
        }
        break;
      }

      const batchDatapoints: DataPoint[] = [];
      for (const datapoint of response.data) {
        const inputs = { ...(datapoint.inputs || {}) };
        if (options?.includeMessages) {
          // Merge top-level messages into inputs if not already present
          if (!("messages" in inputs) && datapoint.messages) {
            inputs.messages = datapoint.messages;
          }
        }
        batchDatapoints.push({
          inputs,
          expectedOutput: datapoint.expectedOutput,
        } as DataPoint);

        startingAfter =
          (datapoint as unknown as { _id?: string })._id ??
          (datapoint as unknown as { id?: string }).id;
      }

      batchNumber++;
      const hasMore = response.hasMore ?? false;

      yield {
        datapoints: batchDatapoints,
        hasMore,
        batchNumber,
      };
      hasYielded = true;

      if (!hasMore) {
        break;
      }
    }
  } catch (error) {
    throw new Error(
      `Failed to fetch dataset ${datasetId}: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}

async function fetchDatasetAsDataPoints(
  orqClient: Orq,
  datasetId: string,
  options?: { includeMessages?: boolean },
): Promise<Promise<DataPoint>[]> {
  const allDatapoints: DataPoint[] = [];
  for await (const batch of fetchDatasetBatches(
    orqClient,
    datasetId,
    options,
  )) {
    allDatapoints.push(...batch.datapoints);
  }
  return allDatapoints.map((dp) => Promise.resolve(dp));
}

/**
 * Check if any evaluator returned pass: false
 */
function checkPassFailures(results: EvaluatorqResult): boolean {
  for (const dataPointResult of results) {
    for (const jobResult of dataPointResult.jobResults || []) {
      for (const evaluatorScore of jobResult.evaluatorScores || []) {
        if (evaluatorScore.score.pass === false) {
          return true;
        }
      }
    }
  }
  return false;
}

/**
 * @param _name - The name of the evaluation run.
 * @param params - The parameters for the evaluation run.
 * @returns The results of the evaluation run.
 */
export async function evaluatorq(
  _name: string,
  params: EvaluatorParams,
): Promise<EvaluatorqResult> {
  const {
    data,
    evaluators = [],
    jobs,
    parallelism = 1,
    print = true,
    description,
  } = params;

  // Initialize tracing if OTEL is configured
  const tracingEnabled = await initTracingIfNeeded();
  const parentContext = tracingEnabled
    ? await captureParentContext()
    : undefined;
  const tracingContext: TracingContext | undefined = tracingEnabled
    ? {
        runId: generateRunId(),
        runName: _name,
        enabled: true,
        parentContext,
      }
    : undefined;

  let orqClient: Orq | undefined;
  const orqApiKey = process.env.ORQ_API_KEY;

  if (orqApiKey) {
    orqClient = await setupOrqClient(orqApiKey);
  }

  const startTime = new Date();

  let datasetId: string | undefined;

  // Handle datasetId case - use streaming fetch
  if ("datasetId" in data) {
    if (!orqApiKey || !orqClient) {
      throw new Error(
        "ORQ_API_KEY environment variable must be set to fetch datapoints from Orq platform.",
      );
    }

    datasetId = data.datasetId;
    const datasetIdValue = datasetId; // Capture for use in callbacks
    const includeMessages = data.includeMessages ?? false;

    // Shared progress state that can be updated from within Effect.promise
    const progressRef = {
      totalDataPoints: 0,
      processedDataPoints: 0,
      phase: "fetching" as "fetching" | "processing",
      done: false,
    };

    // Stream fetch and process batches concurrently
    const streamingProgram = pipe(
      Effect.gen(function* (_) {
        const progress = yield* _(ProgressService);

        // Initialize progress with fetching phase
        yield* _(
          progress.updateProgress({
            totalDataPoints: 0,
            currentDataPoint: 0,
            phase: "fetching",
          }),
        );

        // Start a polling loop to update progress from progressRef
        const progressInterval = setInterval(() => {
          Effect.runPromise(
            progress.updateProgress({
              totalDataPoints: progressRef.totalDataPoints,
              currentDataPoint: progressRef.processedDataPoints,
              phase: progressRef.phase,
            }),
          );
        }, 100);

        // Fetch and process batches with streaming
        let allResults: EvaluatorqResult = [];
        let fetchError: Error | null = null;

        try {
          allResults = yield* _(
            Effect.promise(async () => {
              const results: EvaluatorqResult = [];
              const processingPromises: Promise<EvaluatorqResult>[] = [];
              let datapointIndex = 0;
              let activeCount = 0;

              // Simple semaphore for parallelism control
              const waitForSlot = async () => {
                while (activeCount >= parallelism) {
                  await new Promise((resolve) => setTimeout(resolve, 10));
                }
                activeCount++;
              };

              for await (const batch of fetchDatasetBatches(
                orqClient,
                datasetIdValue,
                { includeMessages },
              )) {
                // Update total as we fetch
                progressRef.totalDataPoints += batch.datapoints.length;
                progressRef.phase = "processing";

                // Start processing this batch immediately
                for (const datapoint of batch.datapoints) {
                  const currentIndex = datapointIndex++;

                  // Wait for available slot
                  await waitForSlot();

                  const promise = (async () => {
                    try {
                      // Process datapoint directly without Effect to avoid progress conflicts
                      const jobResults = await Promise.all(
                        jobs.map(async (job) => {
                          try {
                            // Wrap the job + evaluators in a job span for tracing
                            return await withJobSpan(
                              {
                                runId: tracingContext?.runId || "",
                                rowIndex: currentIndex,
                                parentContext: tracingContext?.parentContext,
                              },
                              async (jobSpan) => {
                                const result = await job(
                                  datapoint,
                                  currentIndex,
                                );
                                setJobNameAttribute(jobSpan, result.name);

                                // Run evaluators for this job
                                const evaluatorScores = await Promise.all(
                                  evaluators.map(async (evaluator) => {
                                    try {
                                      // Wrap evaluator in evaluation span (child of job span)
                                      const score = await withEvaluationSpan(
                                        {
                                          runId: tracingContext?.runId || "",
                                          evaluatorName: evaluator.name,
                                        },
                                        async (evalSpan) => {
                                          const evalResult =
                                            await evaluator.scorer({
                                              data: datapoint,
                                              output: result.output,
                                            });

                                          setEvaluationAttributes(
                                            evalSpan,
                                            evalResult.value,
                                            evalResult.explanation,
                                            evalResult.pass,
                                          );

                                          return evalResult;
                                        },
                                      );
                                      return {
                                        evaluatorName: evaluator.name,
                                        score,
                                      };
                                    } catch (error) {
                                      return {
                                        evaluatorName: evaluator.name,
                                        score: { value: "" as string },
                                        error: error as Error,
                                      };
                                    }
                                  }),
                                );
                                return {
                                  jobName: result.name,
                                  output: result.output,
                                  evaluatorScores,
                                };
                              },
                            );
                          } catch (error) {
                            const err = error as Error & { jobName?: string };
                            return {
                              jobName: err.jobName || "Unknown",
                              output: null,
                              error: error as Error,
                            };
                          }
                        }),
                      );
                      progressRef.processedDataPoints++;
                      return [
                        { dataPoint: datapoint, jobResults },
                      ] as EvaluatorqResult;
                    } finally {
                      activeCount--;
                    }
                  })();
                  processingPromises.push(promise);
                }
              }

              // Wait for all processing to complete
              const resultsNested = await Promise.all(processingPromises);
              for (const resultList of resultsNested) {
                results.push(...resultList);
              }

              return results;
            }),
          );
        } catch (error) {
          fetchError = error as Error;
        } finally {
          // Always stop the progress polling
          clearInterval(progressInterval);
        }

        if (fetchError) {
          throw fetchError;
        }

        // Final progress update with correct counts
        yield* _(
          progress.updateProgress({
            totalDataPoints: progressRef.totalDataPoints,
            currentDataPoint: progressRef.processedDataPoints,
            phase: "processing",
          }),
        );

        return allResults;
      }),
      // Conditionally add table display
      print
        ? Effect.tap((results) => displayResultsTableEffect(results, _name))
        : Effect.tap(() => Effect.void),
      // Send results to Orq when API key is available
      orqApiKey
        ? Effect.tap((results) =>
            sendResultsToOrqEffect(
              orqApiKey,
              _name,
              description,
              datasetId,
              results,
              startTime,
              new Date(),
            ),
          )
        : Effect.tap(() => Effect.void),
      // Provide the progress service
      Effect.provide(ProgressServiceLive),
      // Wrap with progress tracking
      (effect) => withProgress(effect, print),
    );

    // Execute the streaming program
    const streamingResults = await Effect.runPromise(streamingProgram);

    // Shutdown tracing gracefully for streaming case
    if (tracingEnabled) {
      await flushTracing();
      await new Promise((resolve) => setTimeout(resolve, 2000));
      await shutdownTracing();
    }

    // Check for pass failures
    const hasStreamingFailures = checkPassFailures(streamingResults);
    if (hasStreamingFailures) {
      process.exit(1);
    }

    return streamingResults;
  }

  // Non-streaming case: process all data at once
  const dataPromises = data;

  // Execute evaluation with optional tracing
  const executeEvaluation = async (): Promise<EvaluatorqResult> => {
    // Create Effect for processing all data points
    const program = pipe(
      Effect.gen(function* (_) {
        const progress = yield* _(ProgressService);

        // Initialize progress
        yield* _(
          progress.updateProgress({
            totalDataPoints: dataPromises.length,
            currentDataPoint: 0,
            phase: "initializing",
          }),
        );

        // Process data points
        const results = yield* _(
          Effect.forEach(
            dataPromises.map((dataPromise, index) => ({ dataPromise, index })),
            ({ dataPromise, index }) =>
              processDataPointEffect(
                dataPromise instanceof Promise
                  ? dataPromise
                  : Promise.resolve(dataPromise),
                index,
                jobs,
                evaluators,
                parallelism,
                tracingContext,
              ),
            { concurrency: parallelism },
          ),
        );

        return results.flat();
      }),
      // Conditionally add table display
      print
        ? Effect.tap((results) => displayResultsTableEffect(results, _name))
        : Effect.tap(() => Effect.void),
      // Send results to Orq when API key is available
      orqApiKey
        ? Effect.tap((results) =>
            sendResultsToOrqEffect(
              orqApiKey,
              _name,
              description,
              datasetId,
              results,
              startTime,
              new Date(),
            ),
          )
        : Effect.tap(() => Effect.void),
      // Provide the progress service
      Effect.provide(ProgressServiceLive),
      // Wrap with progress tracking
      (effect) => withProgress(effect, print),
    );

    // Run the Effect and convert back to Promise
    return Effect.runPromise(program);
  };

  // Execute evaluation - each job span is independent (no parent evaluation_run span)
  const results = await executeEvaluation();

  // Shutdown tracing gracefully - flush and shutdown before checking failures
  if (tracingEnabled) {
    // Force flush all pending spans
    await flushTracing();
    // Give additional time for network operations
    await new Promise((resolve) => setTimeout(resolve, 2000));
    await shutdownTracing();
  }

  // Check for pass failures and exit if any
  const hasFailures = checkPassFailures(results);
  if (hasFailures) {
    process.exit(1);
  }

  return results;
}

// Create an Effect that runs evaluation and optionally displays results
export const evaluatorqEffect = (
  _name: string,
  params: EvaluatorParams,
): Effect.Effect<EvaluatorqResult, Error, never> => {
  const {
    data,
    evaluators = [],
    jobs,
    parallelism = 1,
    print = true,
    description,
  } = params;

  const startTime = new Date();

  // Handle datasetId case
  if ("datasetId" in data) {
    return Effect.gen(function* (_) {
      const apiKey = process.env.ORQ_API_KEY;
      if (!apiKey) {
        return yield* _(
          Effect.fail(
            new Error(
              "ORQ_API_KEY environment variable must be set to fetch datasets from Orq platform.",
            ),
          ),
        );
      }

      const orqClient = yield* _(
        Effect.tryPromise({
          try: () => setupOrqClient(apiKey),
          catch: (error) =>
            new Error(
              `Failed to setup Orq client: ${error instanceof Error ? error.message : String(error)}`,
            ),
        }),
      );

      if (!orqClient) {
        return yield* _(Effect.fail(new Error("Failed to setup Orq client")));
      }

      const dataPromises = yield* _(
        Effect.tryPromise({
          try: () =>
            fetchDatasetAsDataPoints(orqClient, data.datasetId, {
              includeMessages: data.includeMessages,
            }),
          catch: (error) =>
            error instanceof Error
              ? error
              : new Error(`Failed to fetch dataset: ${String(error)}`),
        }),
      );

      return yield* _(
        runEvaluationEffect(
          dataPromises,
          evaluators,
          jobs,
          parallelism,
          print,
          description,
          _name,
          data.datasetId,
          apiKey,
          startTime,
        ),
      );
    });
  }

  const dataPromises = data;
  return runEvaluationEffect(
    dataPromises,
    evaluators,
    jobs,
    parallelism,
    print,
    description,
    _name,
    undefined,
    undefined,
    startTime,
  );
};

// Extract common evaluation logic
const runEvaluationEffect = (
  dataPromises: (Promise<DataPoint> | DataPoint)[],
  evaluators: EvaluatorParams["evaluators"] = [],
  jobs: Job[],
  parallelism: number,
  print: boolean,
  description: string | undefined,
  evaluationName: string,
  datasetId: string | undefined,
  apiKey: string | undefined,
  startTime: Date,
): Effect.Effect<EvaluatorqResult, Error, never> => {
  // Use API key from parameter or environment
  const orqApiKey = apiKey || process.env.ORQ_API_KEY;

  return pipe(
    Effect.gen(function* (_) {
      const progress = yield* _(ProgressService);

      // Initialize progress
      yield* _(
        progress.updateProgress({
          totalDataPoints: dataPromises.length,
          currentDataPoint: 0,
          phase: "initializing",
        }),
      );

      // Process data points (no tracing in Effect variant for now)
      const results = yield* _(
        Effect.forEach(
          dataPromises.map((dataPromise, index) => ({ dataPromise, index })),
          ({ dataPromise, index }) =>
            processDataPointEffect(
              dataPromise instanceof Promise
                ? dataPromise
                : Promise.resolve(dataPromise),
              index,
              jobs,
              evaluators,
              parallelism,
            ),
          { concurrency: parallelism },
        ),
      );

      return results.flat();
    }),
    // Conditionally add table display
    print
      ? Effect.tap((results) =>
          displayResultsTableEffect(results, evaluationName),
        )
      : Effect.tap(() => Effect.void),
    // Send results to Orq when API key is available
    orqApiKey
      ? Effect.tap((results) =>
          sendResultsToOrqEffect(
            orqApiKey,
            evaluationName,
            description,
            datasetId,
            results,
            startTime,
            new Date(),
          ),
        )
      : Effect.tap(() => Effect.void),
    // Provide the progress service
    Effect.provide(ProgressServiceLive),
    // Wrap with progress tracking
    (effect) => withProgress(effect, print),
  );
};

// Composable evaluatorq with display
export const evaluatorqWithTableEffect = (
  name: string,
  params: EvaluatorParams,
): Effect.Effect<EvaluatorqResult, Error, never> =>
  pipe(
    evaluatorqEffect(name, params),
    Effect.tap((results) => displayResultsTableEffect(results, name)),
  );

File: packages/evaluatorq-py/src/evaluatorq/evaluatorq.py
import asyncio
import os
from collections.abc import Awaitable, Sequence
from datetime import datetime, timezone
from typing import Any, cast

from .fetch_data import fetch_dataset_batches, setup_orq_client
from .processings import process_data_point
from .progress import Phase, ProgressService, with_progress
from .send_results import send_results_to_orq
from .table_display import display_results_table
from .tracing import (
    TracingContext,
    capture_parent_context,
    flush_tracing,
    generate_run_id,
    init_tracing_if_needed,
    shutdown_tracing,
)
from .types import (
    DataPoint,
    DataPointInput,
    DatasetIdInput,
    Evaluator,
    EvaluatorParams,
    EvaluatorqResult,
    Job,
)


def check_pass_failures(results: EvaluatorqResult) -> bool:
    """
    Check if any evaluator returned pass_=False.

    Args:
        results: The evaluation results to check

    Returns:
        True if any evaluator failed (pass_=False), False otherwise
    """
    for data_point_result in results:
        if data_point_result.job_results:
            for job_result in data_point_result.job_results:
                if job_result.evaluator_scores:
                    for evaluator_score in job_result.evaluator_scores:
                        if evaluator_score.score.pass_ is False:
                            return True
    return False


async def evaluatorq(
    name: str,
    params: EvaluatorParams | dict[str, Any] | None = None,
    *,
    data: DatasetIdInput | Sequence[Awaitable[DataPoint] | DataPointInput] | None = None,
    jobs: list[Job] | None = None,
    evaluators: list[Evaluator] | None = None,
    parallelism: int = 1,
    print_results: bool = True,
    description: str | None = None,
    path: str | None = None,
) -> EvaluatorqResult:
    """
    Run an evaluation with the given parameters.

    Can be called with either a params dict/object or keyword arguments:

        # Using keyword arguments (recommended):
        await evaluatorq("name", data=[...], jobs=[...], parallelism=5)

        # Using a dict:
        await evaluatorq("name", {"data": [...], "jobs": [...], "parallelism": 5})

        # Using EvaluatorParams:
        await evaluatorq("name", EvaluatorParams(data=[...], jobs=[...]))

    Args:
        name: Name of the evaluation run
        params: Optional EvaluatorParams instance or dict with all parameters.
        data: The data to evaluate. Either a DatasetIdInput to fetch from Orq platform,
              or a list of DataPoint instances/awaitables.
        jobs: The jobs to run on the data.
        evaluators: The evaluators to use. If not provided, only jobs will run.
        parallelism: Number of jobs to run in parallel. Defaults to 1 (sequential).
        print_results: Whether to print results table to console. Defaults to True.
        description: Optional description for the evaluation run.
        path: Optional path (e.g. "MyProject/MyFolder") to place the experiment
              in a specific project and folder on the Orq platform.

    Returns:
        List of DataPointResult objects

    Raises:
        ValidationError: If parameters fail validation.
        ValueError: If neither params nor required kwargs are provided.
    """
    # Handle params dict/object vs kwargs
    if params is not None:
        # Validate params if passed as dict
        if isinstance(params, dict):
            validated = EvaluatorParams.model_validate(params)
        else:
            validated = params
    elif data is not None and jobs is not None:
        # Use kwargs
        validated = EvaluatorParams(
            data=data,
            jobs=jobs,
            evaluators=evaluators,
            parallelism=parallelism,
            print_results=print_results,
            description=description,
            path=path,
        )
    else:
        raise ValueError(
            "Either 'params' or both 'data' and 'jobs' keyword arguments are required"
        )

    # Extract validated values
    data = validated.data
    jobs = validated.jobs
    evaluators_list = validated.evaluators or []
    parallelism = validated.parallelism
    print_results = validated.print_results
    description = validated.description
    path = validated.path

    # Initialize tracing if OTEL is configured
    tracing_enabled = await init_tracing_if_needed()
    parent_context = await capture_parent_context() if tracing_enabled else None
    tracing_context: TracingContext | None = None
    if tracing_enabled:
        tracing_context = TracingContext(
            run_id=generate_run_id(),
            run_name=name,
            enabled=True,
            parent_context=parent_context,
        )

    orq_api_key = os.environ.get("ORQ_API_KEY")

    start_time = datetime.now(timezone.utc)

    dataset_id: str | None = None

    # Create progress service
    progress = ProgressService()

    # Handle dataset_id case - use streaming fetch
    if isinstance(data, DatasetIdInput):
        orq_client = None

        if orq_api_key:
            orq_client = setup_orq_client(orq_api_key)

        if not orq_api_key or not orq_client:
            raise ValueError(
                "ORQ_API_KEY environment variable must be set to fetch datapoints from Orq platform."
            )
        dataset_id = data.dataset_id
        include_messages = data.include_messages

        # Stream fetch and process batches concurrently
        async def run_streaming_evaluation() -> EvaluatorqResult:
            all_results: EvaluatorqResult = []
            processing_tasks: list[asyncio.Task[list[Any]]] = []
            total_datapoints = 0
            datapoint_index = 0

            # Shared progress state for tracking processed count
            progress_ref = {"processed": 0}

            # Semaphore for controlling parallelism
            data_point_semaphore = asyncio.Semaphore(parallelism)

            async def process_with_semaphore(
                index: int, data_promise: DataPoint
            ) -> list[Any]:
                async with data_point_semaphore:
                    result = await process_data_point(
                        data_promise,
                        index,
                        jobs,
                        evaluators_list,
                        parallelism,
                        None,  # Don't pass progress in streaming mode - use polling instead
                        tracing_context,
                    )
                    progress_ref["processed"] += 1
                    return result

            # Initialize progress with unknown total (streaming mode)
            await progress.update_progress(
                total_data_points=0,
                current_data_point=0,
                phase=Phase.FETCHING,
            )

            # Start a background task to poll and update progress
            stop_polling = False

            async def poll_progress():
                while not stop_polling:
                    await progress.update_progress(
                        total_data_points=total_datapoints,
                        current_data_point=progress_ref["processed"],
                        phase=Phase.PROCESSING
                        if progress_ref["processed"] > 0
                        else Phase.FETCHING,
                    )
                    await asyncio.sleep(0.1)

            polling_task = asyncio.create_task(poll_progress())

            try:
                # Fetch and process batches
                async for batch in fetch_dataset_batches(orq_client, dataset_id, include_messages=include_messages):
                    total_datapoints += len(batch.datapoints)

                    # Start processing this batch immediately
                    for datapoint in batch.datapoints:
                        task = asyncio.create_task(
                            process_with_semaphore(datapoint_index, datapoint)
                        )
                        processing_tasks.append(task)
                        datapoint_index += 1

                # Wait for all processing tasks to complete
                results_nested = await asyncio.gather(*processing_tasks)
            finally:
                # Stop the polling task
                stop_polling = True
                _ = polling_task.cancel()
                try:
                    await polling_task
                except asyncio.CancelledError:
                    pass

            # Final progress update
            await progress.update_progress(
                total_data_points=total_datapoints,
                current_data_point=progress_ref["processed"],
                phase=Phase.PROCESSING,
            )

            # Flatten results
            for result_list in results_nested:
                all_results.extend(result_list)

            return all_results

        results = await with_progress(
            run_streaming_evaluation(), progress, show_progress=print_results
        )

    else:
        # Non-streaming case: process all data at once
        data_promises = cast(list[DataPoint], data)

        async def run_evaluation() -> EvaluatorqResult:
            # Initialize progress
            await progress.update_progress(
                total_data_points=len(data_promises),
                current_data_point=0,
                phase=Phase.INITIALIZING,
            )

            # Process data points with controlled concurrency
            data_point_semaphore = asyncio.Semaphore(max(1, parallelism // len(jobs)))

            async def process_with_semaphore(
                index: int, data_promise: Awaitable[DataPoint] | DataPoint
            ) -> list[Any]:
                async with data_point_semaphore:
                    return await process_data_point(
                        data_promise,
                        index,
                        jobs,
                        evaluators_list,
                        parallelism,
                        progress,
                        tracing_context,
                    )

            tasks = [
                process_with_semaphore(index, data_promise)
                for index, data_promise in enumerate(data_promises)
            ]

            # Gather all results
            results_nested = await asyncio.gather(*tasks)

            # Flatten results
            results: EvaluatorqResult = []
            for result_list in results_nested:
                results.extend(result_list)

            return results

        results = await with_progress(
            run_evaluation(), progress, show_progress=print_results
        )

    # Display results table
    if print_results:
        await display_results_table(results)

    # Upload results to Orq platform if API key is available
    if orq_api_key:
        _ = await send_results_to_orq(
            orq_api_key,
            name,
            description,
            dataset_id,
            results,
            start_time,
            datetime.now(timezone.utc),
            path=path,
        )

    # Shutdown tracing gracefully
    if tracing_enabled:
        await flush_tracing()
        await asyncio.sleep(2)  # Give additional time for network operations
        await shutdown_tracing()

    # Check for pass failures and exit if any
    has_failures = check_pass_failures(results)
    if has_failures:
        import sys

        sys.exit(1)

    return results
